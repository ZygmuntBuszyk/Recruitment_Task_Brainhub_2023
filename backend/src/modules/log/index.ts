import { IRouterDefinition } from '../index';
import { logRouter } from './log.router';
import config from '../../config';

const logRouterDefinition: IRouterDefinition = { path: `${config.API_PREFIX}/log`, router: logRouter };

export default logRouterDefinition;
