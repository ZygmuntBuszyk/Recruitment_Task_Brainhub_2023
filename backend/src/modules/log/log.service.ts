import { celebrate, Joi, Segments } from 'celebrate';
import { DatabaseTable, useStore } from '../../lib/db/datastore';
import * as uuid from 'uuid';

enum LogType {
	WARNING = 'W',
	ERROR = 'E',
	INFO = 'I'
}

export interface ILogRequest {
	name: string;
	email: string;
	log: string;
}

export const relevantSeverinityNumber: number = 50;

export const getRelevantSevernityNumber = (severity: number) => severity >= relevantSeverinityNumber;

export const getRelevantLines = (processedLog: string): string => {
	const processedLogLines: string[] = processedLog.split('\n');

	const errorLines = processedLogLines.filter(line => line.startsWith(LogType.ERROR));

	const filteredErrorLines = errorLines.filter(line => {
		const severity = parseInt(line.split(' ')[2]);

		return getRelevantSevernityNumber(severity);
	});

	return filteredErrorLines.join('\n');
};

export const processTheLog = (log: string) => {
	const filteredLines = getFilteredLines(log);

	const sortedLines = sortByAscendingOrder(filteredLines);

	return sortedLines.join('\n');
};

export const saveLogData = async (logData: ILogRequest): Promise<void> => {
	try {
		console.info(`→ Saving log data to the database`);
		const store = await useStore();

		const table = store.table<ILogRequest>(DatabaseTable.LOG);

		const id = uuid.v4();

		await table.write(id, logData);
		console.info(`→ Log record was successfully saved in the database.} `);
	} catch (e) {
		console.info(`→ Failed to save log data in the database`);
		console.error(e);
		throw e;
	}
};

export const isNonAscii = (line: string) => !/^[\x00-\x7F]*$/.test(line);

export const getFilteredLines = (log: string): string[] => {
	const allLines: string[] = log.split('\n');

	return allLines.filter(line => (Object.values(LogType) as string[]).includes(line[0]) && line[1] === ' ' && !isNonAscii(line));
};

export const sortByAscendingOrder = (filteredLines: string[]): string[] => {
	return filteredLines.sort((a, b) => {
		const aTimeStamp = parseInt(a.split(' ')[1]);
		const bTimeStamp = parseInt(b.split(' ')[1]);
		return aTimeStamp - bTimeStamp;
	});
};

export const logValidationMiddleware = () =>
	celebrate({
		[Segments.BODY]: Joi.object({
			name: Joi.string().required(),
			email: Joi.string().required().email(),
			log: Joi.string().required()
		})
	});

// TODO unit test all of those
export default {
	processTheLog,
	getRelevantLines,
	saveLogData,
	logValidationMiddleware,
	sortByAscendingOrder,
	getFilteredLines,
	isNonAscii,
	getRelevantSevernityNumber
};
