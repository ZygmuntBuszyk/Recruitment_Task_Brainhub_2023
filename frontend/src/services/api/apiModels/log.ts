export enum FormRequestType {
	NAME = 'name',
	EMAIL = 'email',
	LOG = 'log'
}

export interface ILogRequest {
	name: string;
	email: string;
	log: string;
}

export interface ILogResponse {
	data: string;
	status: number;
}
