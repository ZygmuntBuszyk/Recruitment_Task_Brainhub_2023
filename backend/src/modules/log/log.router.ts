import type { Request } from 'express';
import { Router } from 'express';
import { getRelevantLines, ILogRequest, logValidationMiddleware, processTheLog, saveLogData } from './log.service';

const logRouter = Router({});

logRouter.post('/', logValidationMiddleware(), async (req: Request, res) => {
	const { body: logData }: { body: ILogRequest } = req;

	try {
		const processedLog = processTheLog(logData.log);

		await saveLogData({ ...logData, log: processedLog } as ILogRequest);

		const relevantLines = getRelevantLines(processedLog);

		return res.status(200).send(relevantLines);
	} catch (error) {
		return res.status(500).send((error as Error).message || 'Internal Server Error');
	}
});

export { logRouter };
