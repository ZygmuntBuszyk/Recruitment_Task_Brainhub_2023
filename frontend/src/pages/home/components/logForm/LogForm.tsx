import React from 'react';
import styles from './LogForm.module.scss';
import { Button, Input, Form } from 'antd';
import { FormRequestType, ILogRequest } from '@/services/api/apiModels/log';

interface ILogFormProps {
	onSubmit: (values: ILogRequest, callback: Function) => void;
	isLoading?: boolean;
}

function LogForm({ onSubmit, isLoading }: ILogFormProps) {
	const [form] = Form.useForm();

	return (
		<Form
			form={form}
			disabled={isLoading}
			layout='vertical'
			requiredMark={false}
			className={styles['form']}
			onFinish={(values: ILogRequest) => onSubmit(values, form.resetFields)}
		>
			<Form.Item
				className={styles['form__item']}
				label={FormRequestType.NAME.toUpperCase()}
				name={FormRequestType.NAME}
				rules={[
					{
						required: true,
						message: 'Name is required.'
					}
				]}
				hasFeedback
			>
				<Input data-testid='logForm.nameInput' className={styles['form__item-input']} placeholder={'Enter your Name'} />
			</Form.Item>

			<Form.Item
				className={styles['form__item']}
				label={FormRequestType.EMAIL.toUpperCase()}
				name={FormRequestType.EMAIL}
				rules={[
					{
						required: true,
						message: 'Email is required.'
					},
					{
						type: 'email',
						message: 'Email should be valid.'
					}
				]}
				hasFeedback
			>
				<Input data-testid='logForm.emailInput' placeholder={'Enter your email'} />
			</Form.Item>

			<Form.Item
				className={styles['form__item']}
				label={FormRequestType.LOG.toUpperCase()}
				name={FormRequestType.LOG}
				rules={[
					{
						required: true,
						message: 'Log is required.'
					}
				]}
				hasFeedback
			>
				<textarea data-testid='logForm.textarea' placeholder={'Enter your log'} disabled={isLoading} />
			</Form.Item>

			<Form.Item className={styles['form__item']}>
				<Button data-testid='logForm.submitButton' type='primary' htmlType='submit'>
					Process Log
				</Button>
			</Form.Item>
		</Form>
	);
}

export default LogForm;
