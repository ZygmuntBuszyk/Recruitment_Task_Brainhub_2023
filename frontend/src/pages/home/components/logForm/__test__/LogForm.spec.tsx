import React from 'react';
import LogForm from '../LogForm';
import '@testing-library/jest-dom';
import renderer from 'react-test-renderer';
import { fireEvent, render, waitFor } from '@testing-library/react';
import { ILogRequest } from '@/services/api/apiModels/log';

describe('LogForm', function () {
	const DUMMY_TEXT_INPUT = 'dummy input';
	const DUMMY_EMAIL_INPUT = 'dummy@email.com';
	const ONSUBMIT_MOCK = jest.fn().mockImplementation((values: ILogRequest, callback: Function) => {});
	let DUMMY_ISLOADING = false;

	it('renders correctly', () => {
		const component = renderer.create(<LogForm onSubmit={ONSUBMIT_MOCK} isLoading={DUMMY_ISLOADING} />).toJSON();
		expect(component).toMatchSnapshot();
	});

	it('Should be able to submit a form, whenever required inputs are filled', async () => {
		const { getByTestId } = render(<LogForm onSubmit={ONSUBMIT_MOCK} isLoading={DUMMY_ISLOADING} />);
		const nameInput = getByTestId('logForm.nameInput');
		expect(nameInput).toBeTruthy();

		const emailInput = getByTestId('logForm.emailInput');
		expect(emailInput).toBeTruthy();

		const textarea = getByTestId('logForm.textarea');
		expect(textarea).toBeTruthy();

		fireEvent.change(nameInput, { target: { value: DUMMY_TEXT_INPUT } });
		expect((nameInput as HTMLInputElement).value).toBe(DUMMY_TEXT_INPUT);

		fireEvent.change(emailInput, { target: { value: DUMMY_EMAIL_INPUT } });
		expect((emailInput as HTMLInputElement).value).toBe(DUMMY_EMAIL_INPUT);

		fireEvent.change(textarea, { target: { value: DUMMY_TEXT_INPUT } });
		expect((textarea as HTMLInputElement).value).toBe(DUMMY_TEXT_INPUT);

		const registerButton = await getByTestId('logForm.submitButton');
		expect(registerButton).toBeTruthy();

		await waitFor(() => {
			fireEvent.click(registerButton);
		});

		expect(ONSUBMIT_MOCK).toHaveBeenCalledWith(
			{
				name: DUMMY_TEXT_INPUT,
				email: DUMMY_EMAIL_INPUT,
				log: DUMMY_TEXT_INPUT
			} as ILogRequest,
			expect.any(Function)
		);

		expect(ONSUBMIT_MOCK).toBeCalledTimes(1);
	});

	it('Should not be able to submit a form, whenever required inputs arent filled', async () => {
		const { getByTestId } = render(<LogForm onSubmit={ONSUBMIT_MOCK} isLoading={DUMMY_ISLOADING} />);
		const nameInput = getByTestId('logForm.nameInput');
		expect(nameInput).toBeTruthy();

		const emailInput = getByTestId('logForm.emailInput');
		expect(emailInput).toBeTruthy();

		const textarea = getByTestId('logForm.textarea');
		expect(textarea).toBeTruthy();

		fireEvent.change(nameInput, { target: { value: '' } });
		expect((nameInput as HTMLInputElement).value).toBe('');

		fireEvent.change(emailInput, { target: { value: '' } });
		expect((emailInput as HTMLInputElement).value).toBe('');

		fireEvent.change(textarea, { target: { value: '' } });
		expect((textarea as HTMLInputElement).value).toBe('');

		const registerButton = await getByTestId('logForm.submitButton');
		expect(registerButton).toBeTruthy();

		await waitFor(() => {
			fireEvent.click(registerButton);
		});

		expect(ONSUBMIT_MOCK).toBeCalledTimes(0);
	});

	it('Should not be able to submit a form if email is in the wrong format', async () => {
		const { getByTestId } = render(<LogForm onSubmit={ONSUBMIT_MOCK} isLoading={DUMMY_ISLOADING} />);
		const nameInput = getByTestId('logForm.nameInput');
		expect(nameInput).toBeTruthy();

		const emailInput = getByTestId('logForm.emailInput');
		expect(emailInput).toBeTruthy();

		const textarea = getByTestId('logForm.textarea');
		expect(textarea).toBeTruthy();

		fireEvent.change(nameInput, { target: { value: DUMMY_TEXT_INPUT } });
		expect((nameInput as HTMLInputElement).value).toBe(DUMMY_TEXT_INPUT);

		fireEvent.change(emailInput, { target: { value: '' } });
		expect((emailInput as HTMLInputElement).value).toBe('');

		fireEvent.change(textarea, { target: { value: DUMMY_TEXT_INPUT } });
		expect((textarea as HTMLInputElement).value).toBe(DUMMY_TEXT_INPUT);

		const registerButton = await getByTestId('logForm.submitButton');
		expect(registerButton).toBeTruthy();

		await waitFor(() => {
			fireEvent.click(registerButton);
		});

		expect(ONSUBMIT_MOCK).toBeCalledTimes(0);
	});

	it('Should not be able to interact with form whenever form is disabled(submiting)', async () => {
		DUMMY_ISLOADING = true;

		const { getByTestId } = render(<LogForm onSubmit={ONSUBMIT_MOCK} isLoading={DUMMY_ISLOADING} />);
		const nameInput = getByTestId('logForm.nameInput');
		expect(nameInput).toBeTruthy();

		const emailInput = getByTestId('logForm.emailInput');
		expect(emailInput).toBeTruthy();

		const textarea = getByTestId('logForm.textarea');
		expect(textarea).toBeTruthy();

		fireEvent.change(nameInput, { target: { value: DUMMY_TEXT_INPUT } });
		expect((nameInput as HTMLInputElement).value).toBe(DUMMY_TEXT_INPUT);

		fireEvent.change(emailInput, { target: { value: DUMMY_EMAIL_INPUT } });
		expect((emailInput as HTMLInputElement).value).toBe(DUMMY_EMAIL_INPUT);

		fireEvent.change(textarea, { target: { value: DUMMY_EMAIL_INPUT } });
		expect((textarea as HTMLInputElement).value).toBe(DUMMY_EMAIL_INPUT);

		const registerButton = await getByTestId('logForm.submitButton');
		expect(registerButton).toBeTruthy();

		await waitFor(() => {
			fireEvent.click(registerButton);
		});

		expect(ONSUBMIT_MOCK).toBeCalledTimes(0);
	});
});
