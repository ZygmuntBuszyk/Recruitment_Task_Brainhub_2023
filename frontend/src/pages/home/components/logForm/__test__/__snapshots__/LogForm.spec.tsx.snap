// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`LogForm renders correctly 1`] = `
<form
  className="ant-form ant-form-vertical ant-form-hide-required-mark css-dev-only-do-not-override-10ed4xt form"
  onReset={[Function]}
  onSubmit={[Function]}
>
  <div
    className="ant-form-item form__item css-dev-only-do-not-override-10ed4xt"
  >
    <div
      className="ant-row ant-form-item-row css-dev-only-do-not-override-10ed4xt"
      style={{}}
    >
      <div
        className="ant-col ant-form-item-label css-dev-only-do-not-override-10ed4xt"
        style={{}}
      >
        <label
          className="ant-form-item-required"
          htmlFor="name"
          title="NAME"
        >
          NAME
        </label>
      </div>
      <div
        className="ant-col ant-form-item-control css-dev-only-do-not-override-10ed4xt"
        style={{}}
      >
        <div
          className="ant-form-item-control-input"
        >
          <div
            className="ant-form-item-control-input-content"
          >
            <span
              className="ant-input-affix-wrapper form__item-input ant-input-affix-wrapper-has-feedback css-dev-only-do-not-override-10ed4xt"
              onClick={[Function]}
            >
              <input
                aria-required="true"
                className="ant-input css-dev-only-do-not-override-10ed4xt"
                data-testid="logForm.nameInput"
                disabled={false}
                hidden={null}
                id="name"
                onBlur={[Function]}
                onChange={[Function]}
                onFocus={[Function]}
                onKeyDown={[Function]}
                placeholder="Enter your Name"
                style={null}
                type="text"
                value=""
              />
              <span
                className="ant-input-suffix"
              />
            </span>
          </div>
        </div>
      </div>
    </div>
  </div>
  <div
    className="ant-form-item form__item css-dev-only-do-not-override-10ed4xt"
  >
    <div
      className="ant-row ant-form-item-row css-dev-only-do-not-override-10ed4xt"
      style={{}}
    >
      <div
        className="ant-col ant-form-item-label css-dev-only-do-not-override-10ed4xt"
        style={{}}
      >
        <label
          className="ant-form-item-required"
          htmlFor="email"
          title="EMAIL"
        >
          EMAIL
        </label>
      </div>
      <div
        className="ant-col ant-form-item-control css-dev-only-do-not-override-10ed4xt"
        style={{}}
      >
        <div
          className="ant-form-item-control-input"
        >
          <div
            className="ant-form-item-control-input-content"
          >
            <span
              className="ant-input-affix-wrapper ant-input-affix-wrapper-has-feedback css-dev-only-do-not-override-10ed4xt"
              onClick={[Function]}
            >
              <input
                aria-required="true"
                className="ant-input css-dev-only-do-not-override-10ed4xt"
                data-testid="logForm.emailInput"
                disabled={false}
                hidden={null}
                id="email"
                onBlur={[Function]}
                onChange={[Function]}
                onFocus={[Function]}
                onKeyDown={[Function]}
                placeholder="Enter your email"
                style={null}
                type="text"
                value=""
              />
              <span
                className="ant-input-suffix"
              />
            </span>
          </div>
        </div>
      </div>
    </div>
  </div>
  <div
    className="ant-form-item form__item css-dev-only-do-not-override-10ed4xt"
  >
    <div
      className="ant-row ant-form-item-row css-dev-only-do-not-override-10ed4xt"
      style={{}}
    >
      <div
        className="ant-col ant-form-item-label css-dev-only-do-not-override-10ed4xt"
        style={{}}
      >
        <label
          className="ant-form-item-required"
          htmlFor="log"
          title="LOG"
        >
          LOG
        </label>
      </div>
      <div
        className="ant-col ant-form-item-control css-dev-only-do-not-override-10ed4xt"
        style={{}}
      >
        <div
          className="ant-form-item-control-input"
        >
          <div
            className="ant-form-item-control-input-content"
          >
            <textarea
              aria-required="true"
              data-testid="logForm.textarea"
              disabled={false}
              id="log"
              onChange={[Function]}
              placeholder="Enter your log"
            />
          </div>
        </div>
      </div>
    </div>
  </div>
  <div
    className="ant-form-item form__item css-dev-only-do-not-override-10ed4xt"
  >
    <div
      className="ant-row ant-form-item-row css-dev-only-do-not-override-10ed4xt"
      style={{}}
    >
      <div
        className="ant-col ant-form-item-control css-dev-only-do-not-override-10ed4xt"
        style={{}}
      >
        <div
          className="ant-form-item-control-input"
        >
          <div
            className="ant-form-item-control-input-content"
          >
            <button
              className="ant-btn css-dev-only-do-not-override-10ed4xt ant-btn-primary"
              data-testid="logForm.submitButton"
              disabled={false}
              onClick={[Function]}
              type="submit"
            >
              <span>
                Process Log
              </span>
            </button>
          </div>
        </div>
      </div>
    </div>
  </div>
</form>
`;
